### ISSUE TITLE
Cursor moves position after deleting a character and changing line

### ISSUE THREAD

**Issue by sphh** [spyder-ide/spyder#12663]
### Issue Report Checklist

* [x] Searched the [issues page](https://github.com/spyder-ide/spyder/issues?q=is%3Aissue) for similar reports
* [x] Read the relevant sections of the [Spyder Troubleshooting Guide](https://github.com/spyder-ide/spyder/wiki/Troubleshooting-Guide-and-FAQ) and followed its advice
* [x] Reproduced the issue after updating with ``conda update spyder`` (or ``pip``, if not using Anaconda)
* [ ] Could not reproduce inside ``jupyter qtconsole`` (if console-related)
* [ ] Tried basic troubleshooting (if a bug/error)
    * [x] Restarted Spyder
    * [ ] Reset preferences with ``spyder --reset``
    * [ ] Reinstalled the latest version of [Anaconda](https://www.anaconda.com/download/)
    * [x] Tried the other applicable steps from the Troubleshooting Guide
* [x] Completed the **Problem Description**, **Steps to Reproduce** and **Version** sections below

## Problem Description

When replacing a simple string in selected text with `Replace selection`, Spyder takes a long time to replace it and to return control to the user.

### What steps reproduce the problem?

1. Open a new Python file (or use `temp.py` or any existing Python file)
2. Write three lines with `# Spyder is great` (each in its own line)
3. Move the cursor in the first line in front of `great`
4. Press `Del` to delete the character `g` from the word `great`
5. Press the key `Cursor down` once

### What is the expected output? What do you see instead?

I would expect that the cursor stays in the position before the character `g` in the next line, but it moves between `g` and `r`.

If I delete the character `g` with `BACKSPACE`, and move down one line, the cursor stays in the position between `g` and `r`, as expected.

## Versions

* Spyder version: 4.1.2
* Python version: 3.6.9. 64-bit
* Qt version: 5.12.7
* PyQt version: 5.12.3
* Operating System name/version: Linux Mint 19.3 Cinnamon

### Dependencies

```

# Mandatory:
atomicwrites >=1.2.0           :  1.3.0 (OK)
chardet >=2.0.0                :  3.0.4 (OK)
cloudpickle >=0.5.0            :  1.3.0 (OK)
diff_match_patch >=20181111    :  20181111 (OK)
intervaltree                   :  None (OK)
IPython >=4.0                  :  7.13.0 (OK)
jedi =0.15.2                   :  0.15.2 (OK)
keyring                        :  None (OK)
nbconvert >=4.0                :  5.6.1 (OK)
numpydoc >=0.6.0               :  0.9.2 (OK)
parso =0.5.2                   :  0.5.2 (OK)
pexpect >=4.4.0                :  4.8.0 (OK)
pickleshare >=0.4              :  0.7.5 (OK)
psutil >=5.3                   :  5.7.0 (OK)
pygments >=2.0                 :  2.6.1 (OK)
pylint >=0.25                  :  2.4.4 (OK)
pyls >=0.31.9;<0.32.0          :  0.31.9 (OK)
qdarkstyle >=2.8               :  2.8 (OK)
qtawesome >=0.5.7              :  0.7.0 (OK)
qtconsole >=4.6.0              :  4.7.1 (OK)
qtpy >=1.5.0                   :  1.9.0 (OK)
sphinx >=0.6.6                 :  2.4.4 (OK)
spyder_kernels >=1.9.0;<1.10.0 :  1.9.0 (OK)
watchdog                       :  None (OK)
xdg >=0.26                     :  0.26 (OK)
zmq >=17                       :  19.0.0 (OK)

# Optional:
cython >=0.21                  :  None (OK)
matplotlib >=2.0.0             :  None (OK)
numpy >=1.7                    :  1.18.2 (OK)
pandas >=0.13.1                :  None (OK)
scipy >=0.17.0                 :  1.4.1 (OK)
sympy >=0.7.3                  :  None (OK)
```

**Comment by bcolsen** [comment]
I can confirm this, and I agree that it should work as described.

**Comment by juanis2112** [comment]
Thanks @sphh for reporting this and thanks @bcolsen for verifying it. I will take a look.

**Comment by goanpeca** [comment]
I can confirm this issue as well. For Mac devs you can use https://karabiner-elements.pqrs.org/

and map the keys like
<img width="1112" alt="Screen Shot 2020-05-12 at 19 18 14" src="https://user-images.githubusercontent.com/3627835/81757696-57a1bc80-9485-11ea-9d6f-4887fcdadb04.png">

To switch the backspace for a delete

**Comment by bcolsen** [comment]
@goanpeca I think fn+delete one the Mac does delete.

**Comment by goanpeca** [comment]
Indeed :-p did not know that one! Thanks :-)

**Comment by ccordoba12** [comment]
This has been brought up before and we found that it's an (annoying) Qt behavior.

However @juanis2112 has a very good idea on how to make this work as expected.

**Comment by sphh** [comment]
Must have missed it.

At the moment I have a .tex file open in TeXstudio, which is qt based (I have 5.9.5 installed), but that is not written in Python. I don't know, if this information can be helpful in narrowing down the cause …

**Comment by hengin** [comment]
Indeed, this is due to how the QPlainTextEdit widget behaves. Here's a small PyQt5 application where the behaviour can be reproduced:
```
from PyQt5.QtWidgets import QApplication, QPlainTextEdit
app = QApplication([])
widget = QPlainTextEdit()
widget.show()
app.exec_()
```

However, the Spyder CodeEditor widget already overrides keyboard input in
https://github.com/spyder-ide/spyder/blob/9c84cd89625d9ad07f54810588ae6464fa4efef1/spyder/plugins/editor/widgets/codeeditor.py#L4254
so fixing this is easy. It suffices to replace the lines
https://github.com/spyder-ide/spyder/blob/9c84cd89625d9ad07f54810588ae6464fa4efef1/spyder/plugins/editor/widgets/codeeditor.py#L4365
and
https://github.com/spyder-ide/spyder/blob/9c84cd89625d9ad07f54810588ae6464fa4efef1/spyder/plugins/editor/widgets/codeeditor.py#L4386
by `self.stdkey_backspace()`:
https://github.com/spyder-ide/spyder/blob/9c84cd89625d9ad07f54810588ae6464fa4efef1/spyder/plugins/editor/widgets/base.py#L933-L937

However, Spyder also has a similarly annoying behaviour with the "Delete" key that is not reproduced by the PyQt snippet above. I'll try to track down the reason for that before making a PR. But a similar workaround should be possible.

**Comment by dalthviz** [comment]
Thanks @hengin for the help!

**Comment by ccordoba12** [comment]
> However, Spyder also has a similarly annoying behaviour with the "Delete" key that is not reproduced by the PyQt snippet above. I'll try to track down the reason for that before making a PR. But a similar workaround should be possible.

Thanks for your help with that @hengin! @juanis2112 was on the right track to solve this one, but unfortunately got busy with our docs and the community survey.

**Comment by hengin** [comment]
For some reason (platform compatibility?), the Delete-key is handled as a shortcut. The problem disappears if I disable the "delete" shortcut in Tools > Preferences > Keyboard shortcuts.

The responsible piece of code is:
https://github.com/spyder-ide/spyder/blob/9c84cd89625d9ad07f54810588ae6464fa4efef1/spyder/plugins/editor/widgets/codeeditor.py#L2218-L2227

Apparently QPlainTextEdit handles Delete events differently than the above. Indeed, if I manually perform the steps above in a QPlainTextEdit I can reproduce the behaviour. The recording below shows the behaviour of three different methods of character deletion in a QPlainTextEdit (same snippet as I posted above). In each case I press up/down immediately after pressing backspace/delete.
![spyder-delete](https://user-images.githubusercontent.com/17409078/104061820-98f79100-51f9-11eb-9f12-9b2f7a9ae2eb.gif)

Actually, much larger jumps can be produced:
![spyder-extreme-delete](https://user-images.githubusercontent.com/17409078/104062359-9e091000-51fa-11eb-9996-8c8087b7accd.gif)

The reason is most likely that the QPlainTextEdit remembers the rightmost position the cursor was at just before the text is deleted and wants to place the cursor there again. So we just need to clear this memory. The simplest method I've found so far is
```
cursor = self.textCursor()
cursor.setPosition(cursor.position())
self.setTextCursor(cursor)
```

I'll go ahead and make a PR now.

**Comment by ccordoba12** [comment]
Thanks @hengin for your help with this one!

**Comment by hengin** [comment]
Actually, it turns out the Qt-bug is fixed in pyqt5-5.15.2 which makes my workaround (#14559) somewhat redundant. But spyder seems to require pyqt5<5.13 for some reason?

**Comment by ccordoba12** [comment]
Yeah, we haven't had time to increase our requirement to PyQt 5.15, nor we will until Spyder 5, to be released in April (some tests are failing with that version and also the highest version for which Anaconda has packages is 5.12).

So I think we should merge your PR for now. What do you think?

**Comment by hengin** [comment]
Yes, that makes sense.

**Comment by ccordoba12** [comment]
Great! Merging then.

**Comment by sphh** [comment]
A very special thanks to you, @hengin, to track down and solve that rather annoying bug!!!!

**Comment by edbird** [comment]
Found this thread from the search. Can I get a confirmation of the situation as of late Feb 2021?

Has Spyder 5 been released? April 2020 was mentioned above. It's now 2021 and the Debian packages for Debian testing provide version 4.2.1

Just looking for some advice as I can confirm this bug is extremely frustrating. Appreciate it's not anything to do with the Spyder devs and it's a QT issue.

**Comment by sphh** [comment]
@ebird: https://github.com/spyder-ide/spyder/issues/12663#issuecomment-758270969 mentions April 2021, because the comment was written in January 2021.

Yes, it is somewhat annoying, but please keep in mind, that this is a voluntary run project with nobody getting paid, so please be patient. April 2021 is only just 2 months away!

**Comment by edbird** [comment]
> @ebird: [#12663 (comment)](https://github.com/spyder-ide/spyder/issues/12663#issuecomment-758270969) mentions April 2021, because the comment was written in January 2021.
> 
> Yes, it is somewhat annoying, but please keep in mind, that this is a voluntary run project with nobody getting paid, so please be patient. April 2021 is only just 2 months away!

Ok I understand, no problem. Thanks for the update. Fully appreciate the work involved.

**Comment by hengin** [comment]
Note that a workaround has been implemented (see PR #14559) and it is part of the recent Spyder 4.2.2 release: https://github.com/spyder-ide/spyder/releases/tag/v4.2.2

**Comment by edbird** [comment]
> Please note that a workaround has been implemented (see PR #14559) and that it is part of the recent Spyder 4.2.2 release: https://github.com/spyder-ide/spyder/releases/tag/v4.2.2

I've just downloaded it - seems to be working as expected now. Thanks for this, a worth while hack to implement even if you're ditching this for the next major release.
